src/session/content/message_row/media.rs

```diff
@@ -34,159 +37,136 @@ mod imp {
+                        MessageContent::MessageAnimation(data) => {
+                            parse_formatted_text(data.caption)
+                        }
+                        MessageContent::MessagePhoto(data) => parse_formatted_text(data.caption),
+                        _ => unreachable!(),
+                    }
+                }))
+                .bind(&*imp.content, "caption", Some(message));
+            imp.binding.replace(Some(caption_binding));
+
+            // Load media
+            let handler_id =
+                message.connect_content_notify(clone!(@weak self as obj => move |message, _| {
+                    obj.update_media(message);
+                }));
+            imp.handler_id.replace(Some(handler_id));
+            self.update_media(message);
         }
 
-        fn request_mode(&self, _widget: &Self::Type) -> gtk::SizeRequestMode {
-            gtk::SizeRequestMode::HeightForWidth
-        }
+        imp.old_message.set(self.message().as_ref());
     }
-}
-
-glib::wrapper! {
-    pub struct Media(ObjectSubclass<imp::Media>)
-        @extends gtk::Widget;
-}
 
-impl Default for Media {
-    fn default() -> Self {
-        Self::new()
-    }
-}
+    fn update_media(&self, message: &Message) {
+        let imp = self.imp();
 
-impl Media {
-    pub fn new() -> Self {
-        glib::Object::new(&[]).expect("Failed to create Media")
-    }
+        imp.content.set_paintable(None);
 
-    pub fn set_aspect_ratio(&self, aspect_ratio: f64) {
-        self.imp().picture.set_aspect_ratio(aspect_ratio);
-    }
+        let message_content = message.content().0;
+        let file = match message_content {
+            MessageContent::MessageAnimation(ref data) => {
+                imp.content
+                    .set_aspect_ratio(data.animation.width as f64 / data.animation.height as f64);
 
-    pub fn set_paintable(&self, paintable: Option<gdk::Paintable>) {
-        self.imp().picture.set_paintable(paintable);
-    }
+                &data.animation.animation
+            }
+            MessageContent::MessagePhoto(ref data) => {
+                let photo_size = data.photo.sizes.last().unwrap();
+                imp.content
+                    .set_aspect_ratio(photo_size.width as f64 / photo_size.height as f64);
+                &photo_size.photo
+            }
+            _ => unreachable!(),
+        };
 
-    pub fn caption(&self) -> String {
-        self.imp().caption_label.label().into()
+        if file.local.is_downloading_completed {
+            imp.content.set_download_progress(1.0);
+            self.load_media_from_path(&file.local.path);
+        } else {
+            imp.content.set_download_progress(0.0);
+            self.download_media(file.id, &message.chat().session());
+        }
     }
 
-    pub fn set_caption(&self, caption: &str) {
-        if self.caption() == caption {
-            return;
-        }
+    fn download_media(&self, file_id: i32, session: &Session) {
```

Comment:
We should probably have a way to stop the downloading when the message of this widget changes, otherwise we risk to end up with the wrong media to be loaded.

Reply:
We could save the `file_id` before we start the download and in `receiver.attach` we compare it with that `file_id` we get in the closure. Then, we return `glib::Continue(false)` if the `file_id` isn't the same anymore.

Reply:
This actually needs further changes with file downloading in Telegrand, so I'll postpone this issue for now.